### ISSUE TITLE
faild to train keras model( reapet epoch results )

### ISSUE THREAD

**Issue by hamidfathi1998** [keras-team/keras#13846]
hi
 I want to predict the next number and my dataset is [data.xlsx](https://drive.google.com/open?id=1brp9k8wxEOJgg9AGlKh-6koI5A6NZCAm)
train on google colab service( runtime type is GPU )
[train log image ](https://drive.google.com/open?id=1baUuwz8t9M5unRql0fiPBC6xKRy_oo63)

```
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras import regularizers

df = pd.read_excel(r'data.xlsx')

dataset = df.values

Y = dataset[:,0]
X = dataset[:,1]

from sklearn import preprocessing
y = Y.reshape(-1, 1) #returns a numpy array
min_max_scaler = preprocessing.MinMaxScaler()
Y = min_max_scaler.fit_transform(y)

x = X.reshape(-1, 1) #returns a numpy array
min_max_scaler = preprocessing.MinMaxScaler()
X = min_max_scaler.fit_transform(x)

from sklearn.model_selection import train_test_split
X_train, X_val_and_test, Y_train, Y_val_and_test = train_test_split(X, Y, test_size=0.2)
X_val, X_test, Y_val, Y_test = train_test_split(X_val_and_test, Y_val_and_test, test_size=0.2)
print(X_train.shape, X_val.shape, X_test.shape, Y_train.shape, Y_val.shape, Y_test.shape)

model_3 = Sequential([
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01), input_shape=(1,)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1, activation='sigmoid', kernel_regularizer=regularizers.l2(0.01)),
])

model_3.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
hist_3 = model_3.fit(X_train, Y_train,
          batch_size=32,shuffle=False, epochs=10000,
          validation_data=(X_val, Y_val))
```

**Comment by Vincent9797** [comment]
Hi,

2 issues that I found with your code.

First, you should fit `min_max_scaler` only to your training set. After that, you use it to transform your validation and test set.

Secondly, you are trying to predict Y, which is a continuous value. Hence, loss should be defined as `mean_squared_error` or other loss functions suitable for continuous values. A suitable metric for a continuous output could also be mean square error as well.

This is how I would re-write it:
```
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras import regularizers

from sklearn import preprocessing
from sklearn.model_selection import train_test_split

df = pd.read_excel(r'data.xlsx')

dataset = df.values

Y = dataset[:,0]
X = dataset[:,1]


Y = Y.reshape(-1, 1) #returns a numpy array
X = X.reshape(-1, 1) #returns a numpy array

X_train, X_val_and_test, Y_train, Y_val_and_test = train_test_split(X, Y, test_size=0.2)

Y_scaler = preprocessing.MinMaxScaler()
Y_train = Y_scaler.fit_transform(Y_train)

X_scaler = preprocessing.MinMaxScaler()
X_train = X_scaler.fit_transform(X_train)

X_val, X_test, Y_val, Y_test = train_test_split(X_val_and_test, Y_val_and_test, test_size=0.2)

X_val = X_scaler.transform(X_val)
X_test = X_scaler.transform(X_test)
Y_val = Y_scaler.transform(Y_val)
Y_test = Y_scaler.transform(Y_test)

print(X_train.shape, X_val.shape, X_test.shape, Y_train.shape, Y_val.shape, Y_test.shape)

model_3 = Sequential([
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01), input_shape=(1,)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1, activation='linear', kernel_regularizer=regularizers.l2(0.01)),
])

model_3.compile(optimizer='adam',
              loss='mse')
hist_3 = model_3.fit(X_train, Y_train,
          batch_size=32,shuffle=False, epochs=10000,
          validation_data=(X_val, Y_val))
```

**Comment by hamidfathi1998** [comment]
> Hi,
> 
> 2 issues that I found with your code.
> 
> First, you should fit `min_max_scaler` only to your training set. After that, you use it to transform your validation and test set.
> 
> Secondly, you are trying to predict Y, which is a continuous value. Hence, loss should be defined as `mean_squared_error` or other loss functions suitable for continuous values. A suitable metric for a continuous output could also be mean square error as well.
> 
> This is how I would re-write it:
> 
> ```
> import pandas as pd
> from keras.models import Sequential
> from keras.layers import Dense
> from keras.layers import Dropout
> from keras import regularizers
> 
> from sklearn import preprocessing
> from sklearn.model_selection import train_test_split
> 
> df = pd.read_excel(r'data.xlsx')
> 
> dataset = df.values
> 
> Y = dataset[:,0]
> X = dataset[:,1]
> 
> 
> Y = Y.reshape(-1, 1) #returns a numpy array
> X = X.reshape(-1, 1) #returns a numpy array
> 
> X_train, X_val_and_test, Y_train, Y_val_and_test = train_test_split(X, Y, test_size=0.2)
> 
> Y_scaler = preprocessing.MinMaxScaler()
> Y_train = Y_scaler.fit_transform(Y_train)
> 
> X_scaler = preprocessing.MinMaxScaler()
> X_train = X_scaler.fit_transform(X_train)
> 
> X_val, X_test, Y_val, Y_test = train_test_split(X_val_and_test, Y_val_and_test, test_size=0.2)
> 
> X_val = X_scaler.transform(X_val)
> X_test = X_scaler.transform(X_test)
> Y_val = Y_scaler.transform(Y_val)
> Y_test = Y_scaler.transform(Y_test)
> 
> print(X_train.shape, X_val.shape, X_test.shape, Y_train.shape, Y_val.shape, Y_test.shape)
> 
> model_3 = Sequential([
>     Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01), input_shape=(1,)),
>     Dropout(0.3),
>     Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
>     Dropout(0.3),
>     Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
>     Dropout(0.3),
>     Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
>     Dropout(0.3),
>     Dense(1, activation='linear', kernel_regularizer=regularizers.l2(0.01)),
> ])
> 
> model_3.compile(optimizer='adam',
>               loss='mse')
> hist_3 = model_3.fit(X_train, Y_train,
>           batch_size=32,shuffle=False, epochs=10000,
>           validation_data=(X_val, Y_val))
> ```

thanks for your attention.
I change code but the first Epoch to last Epoch is equal you can see  [train log](https://drive.google.com/open?id=1Fwo4xmuxW3cIV0RnAijG2xeEGaIOhmbz) 
so loss and val loss is very less 

and I test my model given just one result 
0.051835876 after max min normalize  given this number 4.453738 for all of dataset

```
import tensorflow as tf
model = tf.keras.models.load_model('number_linear_100.model')
import pandas as pd
df = pd.read_excel(r'C:\Users\Windows\Desktop\test\data.xlsx')

import pandas as pd
df = pd.read_excel(r'C:\Users\Windows\Desktop\test\data.xlsx')


dataset = df.values
acc = 0
iss = 0
listPreLen = 50
for i in range(1,4001,listPreLen):
    Y = dataset[ i : i + listPreLen , 0 ]
    X = dataset[ i : i + listPreLen , 1 : ]
    
    pre = model.predict(X)  
    print(pre[0][0])
# -----------------  results :

0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
0.051835876
---------------
```
why given this result?

**Comment by Vincent9797** [comment]
Take note that RMSE is relative to your Y values. If your Y values are in the range of 0.1, then a RMSE of 0.1 might not be ideal. There is a more detailed explanation here: https://www.kaggle.com/c/competitive-data-science-predict-future-sales/discussion/98629

In addition, I changed the min_max_scaler to a simple normalisation of the data by subtracting the mean and then dividing its standard deviation.

You can look at how I modified your code here:  https://colab.research.google.com/drive/1wEm76N5Xkdp9Sx_jkyN9aEMJjoKCN_BM

**Comment by hamidfathi1998** [comment]
> Take note that RMSE is relative to your Y values. If your Y values are in the range of 0.1, then a RMSE of 0.1 might not be ideal. There is a more detailed explanation here: https://www.kaggle.com/c/competitive-data-science-predict-future-sales/discussion/98629
> 
> In addition, I changed the min_max_scaler to a simple normalisation of the data by subtracting the mean and then dividing its standard deviation.
> 
> You can look at how I modified your code here: https://colab.research.google.com/drive/1wEm76N5Xkdp9Sx_jkyN9aEMJjoKCN_BM

I evaluate your colab code [this](https://colab.research.google.com/drive/1wEm76N5Xkdp9Sx_jkyN9aEMJjoKCN_BM)
but given accuracy: 47.27%
how to better that?

**Comment by sushreebarsa** [comment]
@hamidfathi1998 Moving this issue to closed status as there has been no recent activity, in case you still face the error please create a new issue.Thanks!
